modules/02_Theory.md

```diff
@@ -0,0 +1,371 @@
+
+```ts
+import sayHello from "greetings";
+sayHello("world");
+```
+
+To check this file, the compiler needs to know the type of `sayHello` (is it a function that can accept one string argument?), which opens quite a few additional questions:
+
+1. What kind of module does the module system expect to find at the location where I output this file?
+2. What output JavaScript code should I emit for this input?
+3. When the module system loads that output code, where will it look to find the module specified by `"greetings"`? Will the lookup succeed?
+4. What kind of module is the file resolved by that lookup?
+5. Does the module system allow the kind of module detected in (1) to reference the kind of module detected in (4) with the syntax decided in (2)?
+6. Once the `"greetings"` module has been analyzed, what piece of that module is bound to `sayHello`?
+
+Notice that all of these questions depend on characteristics of the _host_—the system that ultimately consumes the output code to direct its module loading behavior, typically either a runtime (like Node) or bundler (like Webpack). The ECMAScript specification defines how ESM imports and exports match up with each other, but it doesn’t specify how the file lookup in (3), known as _module resolution_, happens, and it doesn’t say anything about other module systems like CommonJS. So runtimes and bundlers, especially those that want to support both ESM and CJS, have a lot of freedom to design their own rules. Consequently, the way TypeScript should answer the questions above can vary dramatically depending on where the code is intended to run. There’s no single right answer, so the compiler must be told the rules through configuration options.
+
+The other key idea to keep in mind is that TypeScript always has to think about these questions in terms of its _output_ JavaScript files, not its _input_ TypeScript (or JavaScript!) files. Over the course of this document, we’ll examine a few concrete examples of where this matters. But for now, we can summarize TypeScript’s job when it comes to modules in those terms:
+
+Understand the **rules of the host** enough
+
+1. to compile files into a valid **output module format**,
+2. to ensure that imports in those **outputs** will **resolve successfully**, and
+3. to know what **type** to assign to **imported names**.
+
+## Who is the host?
+
+Before we move on, it’s worth making sure we’re on the same page about the term _host_, because it will come up frequently. We defined it before as “the system that ultimately consumes the output code to direct its module loading behavior.” In other words, it’s the system outside of TypeScript that TypeScript’s module analysis tries to model:
+
+- When the output code (whether produced by `tsc` or a third-party transpiler) is run directly in a runtime like Node, the runtime is the host.
+- When there is no “output code” because a runtime consumes TypeScript files directly, the runtime is still the host.
```

Comment:
```suggestion
- When there is no “output code” because a runtime consumes TypeScript files directly (e.g. `ts-node`, deno, bun), the runtime is still the host.
```